src/slot.rs

```diff
@@ -1081,79 +1081,66 @@ impl<Conn: Connection> SetWorker<Conn> {
+        // Before we vend out the slot's connection to a client, make sure that
+        // we have space to take it back once they're done with it.
+        let Ok(permit) = self.slot_tx.clone().try_reserve_owned() else {
+            event!(Level::TRACE, "Could not reserve slot_tx permit");
 
-                #[cfg(feature = "probes")]
-                probes::slot__set__claim__failed!(|| (
-                    self.pool_name.as_str(),
-                    id.0,
-                    "No unclaimed slots"
-                ));
-                return Err(Error::NoSlotsReady);
-            };
+            #[cfg(feature = "probes")]
+            probes::slot__set__claim__failed!(|| (
+                self.pool_name.as_str(),
+                id.0,
+                "Could not reserve slot_tx permit; all slots used"
+            ));
 
-            let result = tokio::time::timeout(
-                self.config.health_check_timeout,
-                self.backend_connector.on_acquire(&mut handle),
-            )
-            .await;
+            // This is more of an "all slots in-use" error,
+            // but it should look the same to clients.
+            return Err(Error::NoSlotsReady);
+        };
 
-            let Ok(result) = result else {
-                event!(Level::TRACE, "Timeout performing 'on_acquire' on claim");
-                #[cfg(feature = "probes")]
-                probes::slot__set__claim__failed!(|| (
-                    self.pool_name.as_str(),
-                    id.0,
-                    "Timeout from 'on_acquire'"
-                ));
-                continue;
-            };
-            let Ok(()) = result else {
-                event!(Level::TRACE, "Failed performing 'on_acquire' on claim");
-                #[cfg(feature = "probes")]
-                probes::slot__set__claim__failed!(|| (
-                    self.pool_name.as_str(),
-                    id.0,
-                    "Failed 'on_acquire'"
-                ));
-                continue;
-            };
+        let Some(handle) = self.take_connected_unclaimed_slot(permit, id) else {
+            event!(Level::TRACE, "Failed to take unclaimed slot");
 
             #[cfg(feature = "probes")]
-            probes::slot__set__claim__done!(|| (
+            probes::slot__set__claim__failed!(|| (
                 self.pool_name.as_str(),
                 id.0,
-                handle.slot_id().as_u64()
+                "No unclaimed slots"
             ));
+            return Err(Error::NoSlotsReady);
+        };
+
+        #[cfg(feature = "probes")]
+        probes::slot__set__claim__done!(|| (
+            self.pool_name.as_str(),
+            id.0,
+            handle.slot_id().as_u64()
+        ));
 
-            return Ok(handle);
+        return Ok(handle);
+    }
+
+    // Note that this function is not asynchronous.
+    //
+    // This is intentional: We should not be await-ing in the SetWorker
+    // task when servicing client requests.
```

Comment:
This is great :+1: Removes basically all of my concerns about the task-channel pattern.